Add explorer node plugin system

erikd opened the pull request.

erikd commented on cardano-explorer-node/src/Explorer/Node/Plugin.hs:
```diff
@@ -0,0 +1,41 @@
+    -- Rollback to the specified SlotNumber/HeaderHash.
+  , plugRollbackBlock
+        :: [Trace IO Text -> Point ByronBlock -> IO (Either ExplorerNodeError ())]
+  }
```
I really dislike the type signatures of these two, but `ReaderT SqlBackend` *requires* some form of `*LoggingT`.

dcoutts reviewed (approved):
Looks reasonable to me.

I agree that it's a little subtle that plugin authors typically want an action after insert but before rollback. It makes sense since it's nested. So yes, API docs explaining that would be helpful to people.

erikd committed:
db: Add runDbAction

erikd committed:
node: Add function liftedLogException

Like logException this is useful for debugging because exceptions
thown by part of the db-sync code get caught by the network and
are silently swallowed.

erikd committed:
schema: Add a domain type for the epoch output sum

erikd committed:
db: Basics for the extended db-sync node

erikd committed:
node: Implement the extended db-sync node

coot committed:
Use unit instead of Void in explorer's miniprotocols

erikd committed:
node: Use Async.race_ to ensure proper behavior

Previously, exceptions thrown within 'runDbThread' would be stored
in a 'TVar' to be read from 'Async.wait'. Unfortunately 'wait' was
never called so the exception would never get reported.

After some help with the network protocol types I was able to switch
to using 'race_' which has sane behaviour with regard to exceptions.

erikd committed:
Nix voodoo
